### PR TITLE
Source image name from local yaml file

### PR DESCRIPTION
**Pull request by tofay**
_I saw this discussed on #floki-users in the Metaswitch slack a while back. This is PR is prompted by the fact that twice in the last week I've checked in a change to a floki file or gitlab CI file, and forgot to change the other of the two..._

This adds the ability to get the image name from a local yaml file, which will make it easier to update/synchronize floki files and gitlab CI files.

I wasn't sure how to add tests for this. What would be sufficient here?

I've tested locally with the following yaml file:
```
foo:
  bar:
    - baz:
      ekidd/rust-musl-builder
  baz:
    0: ekidd/rust-musl-builder
```

Specifying the following in floki.yaml works:
```
image:
  yaml:
    file: test.yaml
    key: foo.bar.0.baz
```
as does
```
image:
  yaml:
    file: test.yaml
    key: foo.baz.0
```
There's possibly some more complex edge cases that don't work, but this should cover the vast majority of use cases.

### CODE REVIEW

**Comment by rlupton20** [comment]
Awesome! Thanks for doing this.

You caught me mid bootstrapping my laptop, but I should be able to review this in the next couple of days (and take it for a spin).

Probably this will numb enough pain that it would be worth making a release fairly promptly too.

File: src/errors.rs [hunk 1]
```diff
@@ -53,6 +53,9 @@ pub enum FlokiError {
     #[fail(display = "Failed to check existance of image '{}': {}", image, error)]
     FailedToCheckForImage { image: String, error: io::Error },
 
+    #[fail(display = "Failed to find the specified key")]
```
Comment (rlupton20) [inline src/errors.rs]: We can probably give some more specific diagnostics here (failed to find key 'foo.bar.baz' in file 'config.yaml').

File: src/image.rs [hunk 2]
```diff
@@ -26,14 +35,30 @@ fn default_context() -> String {
...
-            Image::Build { ref build } => build.name.clone() + ":floki",
+            Image::Name(ref s) => Ok(s.clone()),
+            Image::Build { ref build } => Ok(build.name.clone() + ":floki"),
+            Image::Yaml { ref yaml } => {
+                let contents = fs::read_to_string(&yaml.file)?;
+                let raw = YamlLoader::load_from_str(&contents)?;
+                let path = yaml.key.split('.').collect::<Vec<_>>();
+                let mut val = &raw[0];
+                for key in &path {
+                    val = match key.parse::<usize>() {
```
Comment (rlupton20) [inline src/image.rs]: This is somewhat mysterious - I'm a little puzzled as to how this match is working - care to explain?

Comment (rlupton20) [reply src/image.rs]: (probably will become clear on reading the docs)

File: Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,7 @@ serde_derive = "1.0"
 serde_yaml = "0.7"
 structopt = "0.2"
 uuid = { version = "0.6", features = ["v4"] }
+yaml-rust = "0.4.3"
 
 [dev-dependencies]
 tempdir = "0.3.7"
```
Comment (rlupton20) [inline Cargo.toml]: You can probably use `tempdir` for testing this (write out a test yaml file).

A particularly interesting test case is when you have a director structure like

```
- floki.yaml
- config.yaml
- some_dir/
  - blah
```

and run `floki` from `some_dir`. Does it locate the `config.yaml` properly?

File: src/errors.rs (hunk 1 above)
Comment (tofay) [reply src/errors.rs]: ok done

File: src/image.rs (hunk 2 above)
Comment (tofay) [reply src/image.rs]: Added a comment - does that explain it?

File: Cargo.toml (hunk 3 above)
Comment (tofay) [reply Cargo.toml]: No it doesn't locate `config.yaml`! I wasn't aware that floki could be run from a subdirectory...

I've assumed that the image.yaml.file path should be relative to the floki.yaml path (seems reasonable -right?), so I've fixed this by prepending image.yaml.file with the path to floki.yaml's directory.

**Comment by tofay** [comment]
I've done the suggestion markups, and fixed the bug you spotted. I'll sort out the DCO when I'm on my home laptop, and look into adding a test after you've given it a spin and are happy with it.

File: Cargo.toml (hunk 3 above)
Comment (rlupton20) [reply Cargo.toml]: It's a new feature, so it makes sense you wouldn't expect it!

Your assumptions sound reasonable, and align with mine.

File: src/image.rs (hunk 2 above)
Comment (rlupton20) [reply src/image.rs]: I think I just find the API confusing - let me look into it, and I''ll figure out what I want the comment to be. Happy if it works for now :-)

File: src/config.rs [hunk 4]
```diff
@@ -57,13 +57,29 @@ impl FlokiConfig {
...
             }
         })?;
 
+        // Ensure the path to an external yaml file is correct.
+        // If the image.yaml.path file is relative, then it should
+        // be relative to the floki config file. At this point we
+        // already have the path to the floki config file, so we
+        // just prepend that to image.yaml.path.
+        if let image::Image::Yaml { ref mut yaml } = config.image {
+            if yaml.file.is_relative() {
```
Comment (rlupton20) [inline src/config.rs]: Probably the path wants to always be relative, since it probably ought to be part of the same codebase. I guess there is nothing to stop it being absolute though (and maybe that allows some nice abuse).

**Comment by rlupton20** [comment]
Cool - let me take this for a spin tomorrow. I think I'm broadly happy, I can make fixes once this is in master without too much trouble anyway.

If you want to write a couple of basic tests go ahead.

File: src/config.rs [hunk 5]
```diff
@@ -57,13 +57,29 @@ impl FlokiConfig {
...
+                yaml.file = file
+                    .parent()
+                    .ok_or_else(|| errors::FlokiInternalError::InternalAssertionFailed {
+                        description: format!("config_file '{:?}' does not have a parent", &file),
```
Comment (rlupton20) [inline src/config.rs]: Maybe we want a slightly different error message here

**Comment by rlupton20** [comment]
OMG today I learned if you add .diff or .patch to the end of the pull requests URI you get a diff / patch. Different life.

**Comment by rlupton20** [comment]
Tom, this works wonderfully. Thank you! I'm happy to merge. Do you want to address the final comments? Then let's squash this, and get it in.

**Comment by rlupton20** [comment]
(Ideally we'd have some tests also, but we can address that later)

File: src/config.rs (hunk 4 above)
Comment (tofay) [reply src/config.rs]: Yeah I figured no-one would *want* to use an absolute path, but it doesn't hurt to check.

File: src/config.rs (hunk 5 above)
Comment (tofay) [reply src/config.rs]: This is actually unreachable, because we've already done the same check when finding the config file. Shall I change to expect?

Comment (tofay) [reply src/config.rs]: ok, changed this so it makes more sense and no longer matches the other `FlokiInternalError`

**Comment by rlupton20** [comment]
You are happy for me to squash and merge this?

**Comment by rlupton20** [comment]
Ok, just going to go ahead and do it, I'm happy with this.
